Add Port-Forwarding and UPNP tutorial (#7076)
godotengine/godot-docs: tutorials/networking/port_forwarding_upnp.rst

.. _doc_port_forwarding_upnp:

Port Forwarding & Universal Plug n' Play
========================================

Why Universal Plug n' Play?
---------------------------

After you are done testing solo with localhost(127.0.0.1) you want to test on a *real* connection.
You will be likely overwhelmed with your choices:

1. Host a real server and deploy there (deployment of a build will be a chore unless you learn advanced methods like remote ssh deployment which are included in Godot)

2. Peer-to-Peer connection, straight to a computer. This includes NAT Holepunching, STUN/TURN and other methods which are overwhelming compared to :abbr:`UPNP (Universal Plug n Play)`.

> Before any code, let's look into the router setup; Not all Routers support :abbr:`UPNP (Universal Plug n Play)`, nor do all users have :abbr:`UPNP (Universal Plug n Play)` enabled. For this reason it is suggested you use :abbr:`UPNP (Universal Plug n Play)` for testing instead of your final public deployment. Even if you enable your router's :abbr:`UPNP (Universal Plug n Play)` (assuming it is a choice), you need to port-forward the port, because the IP is public but all your router's ports are closed by default.

Author: >Also the real possibility of just hosting locally and forwarding the port manually once (and maybe using DynDNS to deal with changing IPs) should be mentioned, as its probably the easiest way (or hosting yourself if you know what you're doing).

I tried just opening port 9999 (basically the ruleset of this tutorial) and hosting on my public IP, but I couldn't have (local) clients connect to my external public IP. Unless I hosted via UPNP. No idea if its a bug or I did something wrong (probably the latter) but if you can host locally and have others connect to you from a non-local network without UPNP, do share the secret :P

> Before any code, let's look into the router setup; Not all Routers support :abbr:`UPNP (Universal Plug n Play)`, nor do all users have :abbr:`UPNP (Universal Plug n Play)` enabled. For this reason it is suggested you use :abbr:`UPNP (Universal Plug n Play)` for testing instead of your final public deployment. Even if you enable your router's :abbr:`UPNP (Universal Plug n Play)` (assuming it is a choice), you need to port-forward the port, because the IP is public but all your router's ports are closed by default.

Reviewer: You likely need to setup port forwarding on your router to forward the external port 9999 on your public IP to port 9999 on your local internal machine. Basically what UPnP does for you, manually. Not too much of a burden if you only have to do it once for your own game, its a few clicks and doesn't require enabling or messing around with UPnP or adding code for that.

UPnP is (was) mostly useful to do this automatically for less technical users, but as more and more routers come with UPnP disabled due to security concerns, it's growing less and less useful - if you need to teach users how to enable UPnP (if possible at all), you may as well just tell them to create a port forward in the router manually.

How to port-forward
-------------------

> 1. Login to your router (username/password is on the router's bottom where it has barcode tag and original wifi password)

Author: >We should not provide guidance on how to login to your router. That is way too specific, depends on model and vendor and is also just not our scope. :)

I browsed UPNP setups a bit before I found the 2 videos mentioned in OP whereas I got the final UPNP solution mentioned here. Many people who failed (e.g. https://teddit.net/r/godot/comments/121lul7/godot_4_upnp_invalid_gateway), I bet just failed the port forwarding process. 

As for what you said (quoted above), I tested this succesfully on 2 routers. 1 of them did not have UPNP at all, I just port forwarded and it worked with the UPNP code of this tutorial. And by extension, the sub-tutorial of 'How to port forward' applies to basically all routers, and ensures the user will open his ports instead of trying to just enable his router's UPNP, or be discouraged if he finds his router doesn't have UPNP.

>We should just say something to the effect of "make sure UPnP port forwarding is enabled and supported by your router"

Ideally it would be a collapsable section. I don't know how to do that, does sphinx support such an option?

>Also, I don't get the explanation on creating a port-forward; UPnP is used exactly because it avoids having to do that.

Mentioned far above, tl;dr: Enabling UPnP without opening a port isn't enough for 2/2 routers I used, for external clients to connect.

> 1. Login to your router (username/password is on the router's bottom where it has barcode tag and original wifi password)

Author: >If you open and forward a port in your router manually, you don't need UPnP. It will just work. 

It didn't in my case. I will make an MRP when I try again, though that will take a while as it is low priority~

2. Enable UPNP if it exists.
3. Find a section called 'Port Forwarding' or 'Port Mapping'
4. Add new
5. Title/Name: "godot_tcp" (do not capitalize)
6. Local IP Address: 192.168.x.y (x,y depends on your local IP, do find it)
7. Protocol: TCP
8. LAN Port: 9999-9999 (or whichever number you want, just make sure its the same everywhere)
9. Public Port: 9999-9999
10. Apply/Save
11. Add new (this time for UDP)
12. Title/Name: "godot_udp" (do not capitalize)
13. Local IP Address: 192.168.x.y
14. Protocol: UDP
15. LAN Port: 9999-9999
16. Public Port: 9999-9999
17. Apply/Save

UPNP Code
---------------

Now, let's take the following minimal host/client code:

::

    @export var port: int = 9999
    @export var ip: String = "localhost"
    var peer = ENetMultiplayerPeer.new()

    func _enter_tree():
	    multiplayer.peer_connected.connect(add_peer)
	    multiplayer.peer_disconnected.connect(del_peer)

    func start_server():
        peer.create_server(port)
        multiplayer.multiplayer_peer = peer

    func start_client():
        peer.create_client(ip, port)
        multiplayer.multiplayer_peer = peer

> and convert it into UPNP. Reminder that **only the host must use UPNP**. Because the goal is after all to expose a public IP onto the wide internet, accessible by anyone.

Reviewer: ```suggestion
and convert it so it uses UPnP. Reminder that **only the host needs UPnP**. Because the goal is after all to expose the host's port to the wide internet, accessible by anyone.
```

.. _doc_port_forwarding_upnp:

Port Forwarding & Universal Plug and Play
==================================================

Why Universal Plug and Play?
-----------------------------

Besides testing locally on your own, you should test on a *real* network connection to make sure your game actually works under real-life networking conditions. There are multiple ways to achieve this:

>     1. Keep hosting your game locally, but expose it to the public internet. This is easy and quick, but comes with a few downsides. Your public IP is likely dynamically assigned and changes, which requires you to let your testers know each time what to connect to, or to use a dynamic DNS service to point at your changing IP. You also need to deal with any firewalls blocking external access to your local game, most notably on your router, and in the likely case it uses NAT (network address translation), you also need to configure port forwarding on your router. Finally, you open a (small part of) your computer and home entwork to the public internet, which always has some risks.

Reviewer 2: ```suggestion
    1. Keep hosting your game locally, but expose it to the public internet. This is easy and quick, but comes with a few downsides. Your public IP is likely dynamically assigned and changes, which requires you to let your testers know each time what to connect to, or to use a dynamic DNS service to point at your changing IP. You also need to deal with any firewalls blocking external access to your local game, most notably on your router, and in the likely case it uses NAT (network address translation), you also need to configure port forwarding on your router. Finally, you open a (small part of) your computer and home network to the public internet, which always has some risks.
```
